urwid_geventloop.py

```diff
@@ -61,46 +52,38 @@ def remove_alarm(self, handle):
     def _watch_file(self, fd, callback):
         while True:
             select.select([fd], [], [])
-            self._greenlet_spawned(gevent.spawn(callback))
+            gevent.spawn(callback).link(self._greenlet_completed)
 
     def watch_file(self, fd, callback):
         greenlet = gevent.spawn(self._watch_file, fd, callback)
-        return self._greenlet_spawned(greenlet)
+        greenlet.link(self._greenlet_completed)
+        return greenlet
 
     def remove_watch_file(self, handle):
         handle.kill()
         return True
 
     # idle
 
-    def _entering_idle(self):
-        for callback in self._idle_callbacks.values():
-            callback()
-
     def enter_idle(self, callback):
-        self._idle_handle += 1
-        self._idle_callbacks[self._idle_handle] = callback
-        return self._idle_handle
+        self._idle_callbacks.append(callback)
+        return callback
 
     def remove_enter_idle(self, handle):
         try:
-            del self._idle_callbacks[handle]
+            self._idle_callbacks.remove(handle)
```

Comment:
We actually still have a potential race condition here, just like before:  
If `_entering_idle()` is running while we remove a callback, we'll be modifying a list that is being iterated over.